File: Semestr 3/Paradygmaty Programowania - ćwiczenia/lista1/OCaml/lista1_cw.ml
let flatten1 listOfLists = 
  let rec flatten1Helper result remaining =
    if remaining = [] then result
    else 
      let headList, tailLists = List.hd remaining, List.tl remaining in
      let flatList = result @ headList in
      flatten1Helper flatList tailLists
  in

  flatten1Helper [] listOfLists



let rec count x xs = 
  if xs = [] then 0
  else
    let head,tail = List.hd xs, List.tl xs in 
    if head = x then (1+count x tail) 
    else count x tail



let rec replicate x n =
  if n = 0 then []
  else
    x :: replicate x (n-1)



let rec sqrList xs =
  if xs = [] then []
  else
    let head,tail = List.hd xs, List.tl xs in 
    (head * head) :: sqrList tail 



let rec palindrome xs =
  if List.length xs <= 1 then true
  else 
    if (List.hd xs) = (List.nth xs ((List.length xs )- 1)) then
      palindrome (List.tl (List.rev (List.tl xs)))
    else false



let rec listLength xs =
  if xs = [] then 0
  else
    1 + listLength (List.tl xs)



let () =

Printf.printf "\nflatten1:\n";
let listOfLists = [[5; 6]; [1; 2; 3]; [7; 6; 5; 4; 3; 2; 1]] in
Printf.printf "%s\n" (String.concat " " (List.map string_of_int (flatten1 listOfLists)));

Printf.printf "\ncount:\n";
let list1 = [1; 2; 3; 4; 3; 3; 3; 2; 3] in
Printf.printf "%d\n" (count 3 list1);

Printf.printf "\nreplicate:\n";
Printf.printf "%s\n" (String.concat " " (replicate "test" 5));

Printf.printf "\nsqrList:\n";
let list2 = [2; 3; 4; 5] in
Printf.printf "%s\n" (String.concat " " (List.map string_of_int (sqrList list2)));

Printf.printf "\npalindrome:\n";
Printf.printf "%b\n" (palindrome ["a"; "l"; "a"]);
Printf.printf "%b\n" (palindrome ["t"; "e"; "s"; "t"]);
Printf.printf "%b\n" (palindrome ["k"; "a"; "m"; "i"; "l"; "s"; "l"; "i"; "m"; "a"; "k"]);

Printf.printf "\nlistLength:\n";
let list3 = [1; 2; 3; 4; 5] in
Printf.printf "%d\n" (listLength list3);

File: Semestr 3/Paradygmaty Programowania - ćwiczenia/lista01/Scala/lista1_cw.scala
import scala.annotation.tailrec

def flatten1[A](listOfLists: List[List[A]]): List[A] = {
  @tailrec
  def flatten1Helper(result: List[A], remaining: List[List[A]]): List[A] = {
    if (remaining.isEmpty) result
    else {
      val (headList, tailLists) = (remaining.head, remaining.tail)
      val flatList = result ++ headList
      flatten1Helper(flatList, tailLists)
    }
  }
  flatten1Helper(Nil, listOfLists)
}



def count[A](x: A, xs: List[A]): Int = {
  if (xs.isEmpty) 0
  else {
    val (head, tail) = (xs.head, xs.tail)
    if (head == x) 1 + count(x, tail)
    else count(x, tail)
  }
}



def replicate[A] (x: A, n: Int): List[A] = {
  if (n == 0) {
    Nil
  } else {
    x :: replicate(x, n-1)
  }
}



def sqrList(xs: List[Int]): List[Int] = {
  if (xs.isEmpty) {
    Nil
  } else {
    val (head, tail) = (xs.head, xs.tail)
    (head * head) :: sqrList(tail)
  }
}



@tailrec
def palindrome[A](xs: List[A]): Boolean = {
  if (xs.size <= 1) {
    true
  } else if (xs.head == xs.last) {
    palindrome(xs.tail.init)
  } else {
    false
  }
}



def palindrome2[A](xs: List[A]): Boolean = {
  xs == xs.reverse
}



def  listLength[A](xs: List[A]): Int = {
  if (xs.isEmpty) {
    0
  } else {
    1 + listLength(xs.tail)
  }
}



@main
def main(): Unit = {
  println("\nflatten1:")
  val listOfLists = List(List(5, 6), List(1, 2, 3), List(7, 6, 5, 4, 3, 2, 1))
  println(flatten1(listOfLists))

  println("\ncount:")
  val list1 = List(1, 2, 3, 4, 3, 3, 3, 2, 3)
  println(count(3, list1))

  println("\nreplicate:")
  println(replicate("test", 5))

  println("\nsqrList:")
  println(sqrList(List(2, 3, 4, 5)))

  println("\npalindrome:")
  println(palindrome(List("a", "l", "a")))
  println(palindrome(List("t", "e", "s", "t")))
  println(palindrome(List("k", "a", "m", "i", "l", "s", "l", "i", "m", "a", "k")))

  println("\nlistLength:")
  println(listLength(List(1, 2, 3, 4, 5)))
}